modal/volume.py

```diff
@@ -678,9 +727,209 @@ async def _upload_file(self, file_spec: FileUploadSpec) -> api_pb2.MountFile:
+                ]
+                files.append(api_pb2.VolumePutFiles2Request.File(
+                    path=file_spec.path,
+                    mode=file_spec.mode,
+                    size=file_spec.size,
+                    blocks=blocks
+                ))
+
+            request = api_pb2.VolumePutFiles2Request(
+                volume_id=self._volume_id,
+                files=files,
+                disallow_overwrite_existing_files=not self._force,
+            )
+
+            try:
+                response = await retry_transient_errors(self._client.stub.VolumePutFiles2, request, base_delay=1)
+            except GRPCError as exc:
+                raise FileExistsError(exc.message) if exc.status == Status.ALREADY_EXISTS else exc
+
+            if not response.missing_blocks:
+                break
+
+            # TODO(dflemstr): call `self._progress_cb` based on `num_blocks_total - len(response.missing_blocks)`
+            # self._progress_cb()
+
+            await _put_missing_blocks(file_specs, response.missing_blocks, put_responses)
+        else:
+            raise RuntimeError("Did not succeed at uploading all files despite supplying all missing blocks")
+
+        self._progress_cb(complete=True)
+
+
+VolumeUploadContextManager2 = synchronize_api(_VolumeUploadContextManager2)
+
+
+async def _put_missing_blocks(
+    file_specs: list[FileUploadSpec2],
+    # TODO(dflemstr): Element type is `api_pb2.VolumePutFiles2Response.MissingBlock` but synchronicity gets confused
+    # by the nested class (?)
+    missing_blocks: list,
+    put_responses: dict[bytes, bytes]
+):
+    # Lazy load slow dep only if this code path is triggered
+    import aiohttp
+
+    async with aiohttp.ClientSession() as session:
+        async def put_missing_block(
+            # TODO(dflemstr): Type is `api_pb2.VolumePutFiles2Response.MissingBlock` but synchronicity gets confused
+            # by the nested class (?)
+            missing_block
+        ) -> (bytes, bytes):
+            assert isinstance(missing_block, api_pb2.VolumePutFiles2Response.MissingBlock)
+
+            missing_block_file_spec = file_specs[missing_block.file_index]
+            # TODO(dflemstr): What if the underlying file has changed here in the meantime; should we check the
+            #  hash again just to be sure?
+            missing_block_sha256 = missing_block_file_spec.blocks_sha256[missing_block.block_index]
+            missing_block_start = missing_block.block_index * BLOCK_SIZE
+
+            with missing_block_file_spec.source() as source_fp:
+                data = b""
+                num_bytes_read = 0
+
+                source_fp.seek(missing_block_start)
+
+                while num_bytes_read < BLOCK_SIZE:
+                    # Most of the time, this will be a single read() call, so don't optimize too much for this
+                    # loop having more than one iteration.
+                    chunk = source_fp.read(BLOCK_SIZE - num_bytes_read)
```

Comment:
Not sure how much we need to worry about this read being blocking. I suppose in the general case, lets say you are reading from a network drive, it could potentially be bad since it would lock up the event loop, but for really small reads on local ssd it feels like putting this in a to_thread might make things actually go slower. wdyt?